### PR TITLE
Allow for configuring MinWatchTimeout in reflector

### PR DESCRIPTION
**Pull request by wojtek-t**
The third commit bump it from default 5m to 30m for kubelet secret/configmap watches to show it's usable, but we can move it to a separate PR.

The motivation for that change is that every watch recreation comes with some additional cost and with watch being much more robust than in early days of kubernetes, we have an opportunity to reduce that cost and increase scalability by simply making the watches to last longer...

Notes to reviewer: I suggest reviewing commit by commit.

```release-note
NONE
```

/sig api-machinery
/kind cleanup
/priority important-soon

### CODE REVIEW

**Comment by wojtek-t** [comment]
/assign @jpbetz 

Joe - would you mind taking a look?

File: pkg/kubelet/util/manager/watch_based_manager.go [hunk 1]
```diff
@@ -224,12 +224,16 @@ func (c *objectCache) newReflectorLocked(namespace, name string) *objectCacheIte
...
 		&cache.ListWatch{ListFunc: listFunc, WatchFunc: watchFunc},
 		c.newObject(),
 		store,
-		0,
+		cache.ReflectorOptions{
+			Name: fmt.Sprintf("object-%q/%q", namespace, name),
+			// Bump default 5m MinWatchTimeout to avoid recreating
+			// watches too often.
+			MinWatchTimeout: 30 * time.Minute,
+		},
```
Comment (jpbetz) [inline pkg/kubelet/util/manager/watch_based_manager.go:236]: I noticed we don't have any tests in this PR.  I don't know if this kubelet code can tested, but maybe unit test the options to ensure that MinWatchTimeout default is working as expected?

File: staging/src/k8s.io/client-go/tools/cache/controller.go [hunk 2]
```diff
@@ -346,6 +346,49 @@ func DeletionHandlingObjectToName(obj interface{}) (ObjectName, error) {
...
+	ResyncPeriod time.Duration
+
+	// Indexers, if set, are the indexers for the received objects to optimize
+	// certain queries.
+	Indexers Indexers
+
+	// Transform function, if set, will be called on all objects before they will be
+	// put into the Store and corresponding Add/Modify/Delete handlers will be invoked
+	// for them.
+	Transform TransformFunc
```
[the remark this answers is not in the document]
Comment (wojtek-t) [reply staging/src/k8s.io/client-go/tools/cache/controller.go]: Tried to clarify - PTAL

File: staging/src/k8s.io/client-go/tools/cache/controller.go [hunk 3]
```diff
@@ -59,6 +59,11 @@ type Config struct {
 	// FullResyncPeriod is the period at which ShouldResync is considered.
 	FullResyncPeriod time.Duration
 
+	// MinWatchTimeout, if set, will define the minimum timeout for watch requests send
+	// to kube-apiserver. However, values lower than 5m will not be honored to avoid
+	// negative performance impact on controlplane.
+	MinWatchTimeout time.Duration
```
[the remark this answers is not in the document]
Comment (wojtek-t) [reply staging/src/k8s.io/client-go/tools/cache/controller.go]: done

File: pkg/kubelet/util/manager/watch_based_manager.go (hunk 1 above)
Comment (wojtek-t) [reply pkg/kubelet/util/manager/watch_based_manager.go:236]: testing kubelet here would be tricky - I added a unit test of the reflector itself PTAL

**Review by wojtek-t — COMMENTED** [review]
thanks for the review - PTAL

**Comment by jpbetz** [comment]
/approve
/lgtm

This looks great

**Pull request merged**